coral-hive/src/main/java/com/linkedin/coral/transformers/OneBasedArrayIndexTransformer.java

```diff
@@ -0,0 +1,57 @@
+/**
+ * Copyright 2022-2023 LinkedIn Corporation. All rights reserved.
+ * Licensed under the BSD-2 Clause license.
+ * See LICENSE in the project root for license information.
+ */
+package com.linkedin.coral.transformers;
+
+import java.math.BigDecimal;
+
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlNumericLiteral;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.type.ArraySqlType;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlValidator;
+
+import com.linkedin.coral.common.transformers.OperatorTransformer;
+
+
+/**
+ * Transformer to convert SqlCall from array[i] to array[i+1] to ensure array indexes start at 1.
+ */
+public class OneBasedArrayIndexTransformer extends OperatorTransformer {
+
+  private static final String ITEM_OPERATOR = "ITEM";
+
+  public OneBasedArrayIndexTransformer(SqlValidator sqlValidator) {
+    super(sqlValidator);
+  }
+
+  @Override
+  public boolean condition(SqlCall sqlCall) {
+    if (ITEM_OPERATOR.equalsIgnoreCase(sqlCall.getOperator().getName())) {
+      final SqlNode columnNode = sqlCall.getOperandList().get(0);
+      return getRelDataType(columnNode) instanceof ArraySqlType;
```

Comment:
Do you agree this has an assumption that the caller is coming from a SqlShuttle? What are the conditions on the SqlNode that can be passed here? Seems it is restricted (I think must be at the same SELECT as the condition's sqlCall). Is there an additional requirement that it must be a child of it?

Reply:
Yes, the caller is coming from a SqlShuttle, but that is the case for all the SqlNode converters like `CoralSqlNodeToSparkSqlNodeConverter`. 
The condition is that the parent SqlSelect should be visited before we call this method on a child sqlNode, but it's still the case for the SqlNode traversal pattern. I don't think we need to derive the type of a SqlNode without visiting its parent SqlSelect in advance. (If we don't visit the parent SqlSelect first, we can't even transform the child SqlNode of it)